File: packages/sdk-js/src/verify.ts
import { canonicalize, PassportSchema, EnvelopeSchema, type Passport, type Envelope } from "@openpassport/spec";
import { base64UrlToBytes, verify } from "./keys.js";
import { fetchPassport } from "./fetch.js";

export interface VerifyPassportResult {
  valid: boolean;
  passport: Passport | null;
  errors: string[];
  warnings: string[];
}

export async function verifyPassport(input: unknown): Promise<VerifyPassportResult> {
  const errors: string[] = [];
  const warnings: string[] = [];

  // 1. Schema validation
  const parsed = PassportSchema.safeParse(input);
  if (!parsed.success) {
    return {
      valid: false,
      passport: null,
      errors: parsed.error.issues.map((i) => `${i.path.join(".")}: ${i.message}`),
      warnings: [],
    };
  }

  const passport = parsed.data;

  // 2. Verify signature
  try {
    const publicKey = base64UrlToBytes(passport.publicKey);
    const signature = base64UrlToBytes(passport.signature);
    const canonical = canonicalize(passport as unknown as Record<string, unknown>);
    const data = new TextEncoder().encode(canonical);

    const isValid = await verify(signature, data, publicKey);
    if (!isValid) {
      errors.push("Signature verification failed: signature does not match public key");
    }
  } catch (err) {
    errors.push(`Signature verification error: ${err instanceof Error ? err.message : String(err)}`);
  }

  // 3. Check expiration
  const now = new Date();
  const expiresAt = new Date(passport.expiresAt);
  if (expiresAt <= now) {
    errors.push(`Passport expired at ${passport.expiresAt}`);
  }

  // 4. Check issuedAt
  const issuedAt = new Date(passport.issuedAt);
  if (issuedAt > now) {
    warnings.push(`Passport issuedAt is in the future: ${passport.issuedAt} (possible clock skew)`);
  }

  return {
    valid: errors.length === 0,
    passport,
    errors,
    warnings,
  };
}

export interface VerifyMessageResult {
  valid: boolean;
  envelope: Envelope | null;
  passport: Passport | null;
  errors: string[];
}

export async function verifyMessage(
  input: unknown,
  options?: { fetchPassport?: boolean; passport?: Passport; maxAgeSeconds?: number },
): Promise<VerifyMessageResult> {
  const errors: string[] = [];

  // 1. Schema validation
  const parsed = EnvelopeSchema.safeParse(input);
  if (!parsed.success) {
    return {
      valid: false,
      envelope: null,
      passport: null,
      errors: parsed.error.issues.map((i) => `${i.path.join(".")}: ${i.message}`),
    };
  }

  const envelope = parsed.data;

  // 2. Get passport
  let passport: Passport | null = options?.passport ?? null;

  if (!passport && options?.fetchPassport !== false) {
    try {
      passport = await fetchPassport(envelope.passportUrl);
    } catch (err) {
      return {
        valid: false,
        envelope,
        passport: null,
        errors: [`Could not fetch passport from ${envelope.passportUrl}: ${err instanceof Error ? err.message : String(err)}`],
      };
    }
  }

  if (!passport) {
    return {
      valid: false,
      envelope,
      passport: null,
      errors: ["No passport provided and fetchPassport is disabled"],
    };
  }

  // 3. Verify passport itself
  const passportResult = await verifyPassport(passport);
  if (!passportResult.valid) {
    return {
      valid: false,
      envelope,
      passport,
      errors: passportResult.errors.map((e) => `Passport: ${e}`),
    };
  }

  // 4. Check from matches passport id
  if (envelope.from !== passport.id) {
    errors.push(`Envelope 'from' (${envelope.from}) does not match passport id (${passport.id})`);
  }

  // 5. Verify envelope signature
  try {
    const publicKey = base64UrlToBytes(passport.publicKey);
    const signature = base64UrlToBytes(envelope.signature);
    const canonical = canonicalize(envelope as unknown as Record<string, unknown>);
    const data = new TextEncoder().encode(canonical);

    const isValid = await verify(signature, data, publicKey);
    if (!isValid) {
      errors.push("Message signature verification failed");
    }
  } catch (err) {
    errors.push(`Message signature error: ${err instanceof Error ? err.message : String(err)}`);
  }

  // 6. Check timestamp freshness
  const maxAge = options?.maxAgeSeconds ?? 300; // 5 minutes
  const msgTime = new Date(envelope.timestamp);
  const age = (Date.now() - msgTime.getTime()) / 1000;
  if (age > maxAge) {
    errors.push(`Message is ${Math.round(age)}s old, exceeds max age of ${maxAge}s`);
  }
  if (age < -30) {
    errors.push(`Message timestamp is in the future by ${Math.round(-age)}s`);
  }

  return {
    valid: errors.length === 0,
    envelope,
    passport,
    errors,
  };
}

File: packages/sdk-python/openpassport/verify.py
"""Verify OpenPassport passports and message envelopes."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from openpassport.canonical import canonicalize
from openpassport.keys import base64url_to_bytes, verify_signature


@dataclass
class VerifyResult:
    valid: bool
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


def verify_passport(passport: dict[str, Any]) -> VerifyResult:
    """Verify a passport document's signature and validity."""
    errors: list[str] = []
    warnings: list[str] = []

    # Check required fields
    required = ["openpassport", "id", "name", "issuer", "publicKey", "endpoint",
                "capabilities", "issuedAt", "expiresAt", "signature"]
    for f in required:
        if f not in passport:
            errors.append(f"Missing required field: {f}")
    if errors:
        return VerifyResult(valid=False, errors=errors)

    # Check spec version
    if passport["openpassport"] != "0.1.0":
        errors.append(f"Unsupported spec version: {passport['openpassport']}")

    # Check ID format
    if not passport["id"].startswith("ap_"):
        errors.append(f"Invalid ID format: {passport['id']}")

    # Verify signature
    try:
        public_key = base64url_to_bytes(passport["publicKey"])
        signature = base64url_to_bytes(passport["signature"])
        canonical = canonicalize(passport)
        data = canonical.encode("utf-8")

        if not verify_signature(signature, data, public_key):
            errors.append("Signature verification failed: signature does not match public key")
    except Exception as e:
        errors.append(f"Signature verification error: {e}")

    # Check expiration
    now = datetime.now(timezone.utc)
    try:
        expires_at = datetime.fromisoformat(passport["expiresAt"].replace("Z", "+00:00"))
        if expires_at <= now:
            errors.append(f"Passport expired at {passport['expiresAt']}")
    except ValueError:
        errors.append(f"Invalid expiresAt format: {passport['expiresAt']}")

    # Check issuedAt
    try:
        issued_at = datetime.fromisoformat(passport["issuedAt"].replace("Z", "+00:00"))
        if issued_at > now:
            warnings.append(f"Passport issuedAt is in the future: {passport['issuedAt']}")
    except ValueError:
        pass

    return VerifyResult(valid=len(errors) == 0, errors=errors, warnings=warnings)


def verify_message(
    envelope: dict[str, Any],
    passport: dict[str, Any],
    max_age_seconds: int = 300,
) -> VerifyResult:
    """Verify a signed message envelope against a passport."""
    errors: list[str] = []

    # Check required fields
    required = ["openpassport_message", "from", "passportUrl", "timestamp", "nonce", "body", "signature"]
    for f in required:
        if f not in envelope:
            errors.append(f"Missing required field: {f}")
    if errors:
        return VerifyResult(valid=False, errors=errors)

    # Verify the passport itself
    passport_result = verify_passport(passport)
    if not passport_result.valid:
        return VerifyResult(
            valid=False,
            errors=[f"Passport: {e}" for e in passport_result.errors],
        )

    # Check from matches passport id
    if envelope["from"] != passport["id"]:
        errors.append(f"Envelope 'from' ({envelope['from']}) does not match passport id ({passport['id']})")

    # Verify envelope signature
    try:
        public_key = base64url_to_bytes(passport["publicKey"])
        signature = base64url_to_bytes(envelope["signature"])
        canonical = canonicalize(envelope)
        data = canonical.encode("utf-8")

        if not verify_signature(signature, data, public_key):
            errors.append("Message signature verification failed")
    except Exception as e:
        errors.append(f"Message signature error: {e}")

    # Check timestamp freshness
    try:
        msg_time = datetime.fromisoformat(envelope["timestamp"].replace("Z", "+00:00"))
        now = datetime.now(timezone.utc)
        age = (now - msg_time).total_seconds()
        if age > max_age_seconds:
            errors.append(f"Message is {int(age)}s old, exceeds max age of {max_age_seconds}s")
        if age < -30:
            errors.append(f"Message timestamp is in the future by {int(-age)}s")
    except ValueError:
        errors.append(f"Invalid timestamp format: {envelope['timestamp']}")

    return VerifyResult(valid=len(errors) == 0, errors=errors)
